docs/modules/ROOT/pages/servlet/authentication/adaptive.adoc

```diff
@@ -0,0 +1,528 @@
+        .formLogin(Customizer.withDefaults())
+        .oneTimeTokenLogin(Customizer.withDefaults());
+    // @formatter:on
+    return http.build();
+}
+----
+
+Kotlin::
++
+[source,kotlin,role="secondary"]
+----
+@Bean
+public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+    // @formatter:off
+    http {
+        authorizeHttpRequests {
+            authorize(anyRequest, authenticated)
+        }
+        formLogin { }
+        oneTimeTokenLogin { }
+    }
+    // @formatter:on
+    return http.build()
+}
+----
+======
+
+By default, this application has two authentication mechanisms that it allows, meaning that the user could use either one and be fully-authenticated.
+
+If there is a set of endpoints that require a specific factor, we can specify that in `authorizeHttpRequests` as follows:
+
+[tabs]
+======
+Java::
++
+[source,java,role="primary"]
+----
+@Bean
+public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+    // @formatter:off
+    http
+        .authorizeHttpRequests((authorize) -> authorize
+            .requestMatchers("/profile/**").hasAuthority("FACTOR_OTT") <1>
+            .anyRequest().authenticated()
+        )
+        .formLogin(Customizer.withDefaults())
+        .oneTimeTokenLogin(Customizer.withDefaults());
+    // @formatter:on
+    return http.build();
+}
+----
+
+Kotlin::
++
+[source,kotlin,role="secondary"]
+----
+@Bean
+public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+    // @formatter:off
+    http {
+        authorizeHttpRequests {
+            authorize("/profile/**", hasAuthority("FACTOR_OTT")) <1>
+            authorize(anyRequest, authenticated)
+        }
+        formLogin { }
+        oneTimeTokenLogin { }
+    }
+    // @formatter:on
+    return http.build()
+}
+----
+======
+<1> - States that all `/profile/**` endpoints require one-time-token login to be authorized
+
+Given the above configuration, users can log in with any mechanism that you support.
+And, if they want to visit the profile page, then Spring Security will redirect them to the One-Time-Token Login page to obtain it.
+
+In this way, the authority given to a user is directly proportional to the amount of proof given.
+This adaptive approach allows users to give only the proof needed to perform their intended operations.
+
+== Multi-Factor Authentication
+
+You may require that all users require both One-Time-Token login and Username/Password login to access any part of your site.
+
+To require both, you can state an authorization rule with `anyRequest` like so:
+
+[tabs]
+======
+Java::
++
+[source,java,role="primary"]
+----
+@Bean
+public SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
+    // @formatter:off
+    http
+        .authorizeHttpRequests((authorize) -> authorize
+            .anyRequest().access(allOf(hasAuthority("FACTOR_PASSWORD"), hasAuthority("FACTOR_OTT"))) <1>
```

Comment:
```suggestion
            .anyRequest().hasAllAuthorities("FACTOR_PASSWORD", "FACTOR_OTT") <1>
```

This can now use https://github.com/spring-projects/spring-security/issues/17932